src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -894,101 +934,326 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
+        /// </summary>
+        private static uint GetInvocationEscapeScope(
+            BoundExpression receiverOpt,
+            ImmutableArray<ParameterSymbol> parameters,
+            ImmutableArray<BoundExpression> args,
+            ImmutableArray<RefKind> argRefKinds,
+            ImmutableArray<int> argsToParamsOpt,
+            uint scopeOfTheContainingExpression)
+        {
+            // SPEC: (also applies to the CheckInvocationEscape counterpart)
+            //
+            //            An lvalue resulting from a ref-returning method invocation e1.M(e2, ...) is ref-safe - to - escape the smallest of the following scopes:
+            //•	The entire enclosing method
+            //•	the ref-safe-to-escape of all ref/out/in argument expressions(excluding the receiver and arguments that have ref-like type)
+            //•	the safe-to - escape of all argument expressions(including the receiver)
+            //
+            //            An rvalue resulting from a method invocation e1.M(e2, ...) is safe - to - escape from the smallest of the following scopes:
+            //•	The entire enclosing method
+            //•	the ref-safe-to-escape of all ref/out/in argument expressions(excluding the receiver and arguments that have ref-like type)
+            //•	the safe-to-escape of all argument expressions(including the receiver)
+            //
+            // Note that these rules are identical to the above rules for ref-safe - to - escape, but apply only when the return type is a ref struct type.
+
+            //by default it is safe to escape
+            uint escapeScope = Binder.ExternalScope;
+
+            ArrayBuilder<bool> inParametersMatchedWithArgs = null;
+
+            if (!args.IsDefault)
             {
-                var indexerAccess = expr as BoundIndexerAccess;
-                if (indexerAccess != null)
+                for (var argIndex = 0; argIndex < args.Length; argIndex++)
                 {
-                    var indexer = indexerAccess.Indexer;
+                    RefKind effectiveRefKind = GetEffectiveRefKind(argIndex, argRefKinds, parameters, argsToParamsOpt, ref inParametersMatchedWithArgs);
 
-                    return CheckArgumentsReturnable(
-                        propertySyntax,
-                        indexer,
-                        indexer.Parameters,
-                        indexerAccess.Arguments,
-                        indexerAccess.ArgumentRefKindsOpt,
-                        indexerAccess.ArgsToParamsOpt,
-                        checkingReceiver,
-                        diagnostics);
+                    // if ref is passed and it is not a ref-like, check ref escape, 
+                    // since it is the same or narrower than val escape and callee may wrap the ref into a val
+                    var argument = args[argIndex];
+                    var argEscape = effectiveRefKind != RefKind.None && argument.Type.IsByRefLikeType == false ?
```

Comment:
I see that we handle the case of null receiver Type (below). Could the argument Type be null too?

Reply:
I am a bit paranoid here with `.Type` it is unusual for the type to be null unless we have errors, but it does happen and sometimes we get to analyze broken code. I just want to be tolerant to situations when it happens intentionally or not, since I generally have a default behavior.
I will add `?.` here, although it is hard to tell whether it may happen or not.

---
In reply to: [135897094](https://github.com/dotnet/roslyn/pull/21334#discussion_r135897094) [](ancestors = 135897094)